hcl2template/types.required_plugins.go

```diff
@@ -0,0 +1,231 @@
+package hcl2template
+
+import (
+	"fmt"
+
+	"github.com/hashicorp/go-version"
+	"github.com/hashicorp/hcl/v2"
+	"github.com/hashicorp/packer/hcl2template/addrs"
+	"github.com/zclconf/go-cty/cty"
+)
+
+func (cfg *PackerConfig) decodeRequiredPluginsBlock(f *hcl.File) hcl.Diagnostics {
+	var diags hcl.Diagnostics
+
+	content, moreDiags := f.Body.Content(configSchema)
+	diags = append(diags, moreDiags...)
+
+	for _, block := range content.Blocks {
+		switch block.Type {
+		case packerLabel:
+			content, contentDiags := block.Body.Content(packerBlockSchema)
+			diags = append(diags, contentDiags...)
+
+			// We ignore "packer_version"" here because
+			// sniffCoreVersionRequirements already dealt with that
+
+			for _, innerBlock := range content.Blocks {
+				switch innerBlock.Type {
+				case "required_plugins":
+					reqs, reqsDiags := decodeRequiredPluginsBlock(innerBlock)
+					diags = append(diags, reqsDiags...)
+					cfg.Packer.RequiredPlugins = append(cfg.Packer.RequiredPlugins, reqs)
+				default:
+					continue
+				}
+
+			}
+		}
+	}
+	return diags
+}
+
+func (cfg *PackerConfig) decodeImplicitRequiredPluginsBlocks(f *hcl.File) hcl.Diagnostics {
+	// when a plugin is used but not defined in the required plugin blocks, it
+	// is 'implicitly used'. Here we read common configuration blocks to try to
+	// guess plugins.
+
+	var diags hcl.Diagnostics
+
+	content, moreDiags := f.Body.Content(configSchema)
+	diags = append(diags, moreDiags...)
+
+	for _, block := range content.Blocks {
+		switch block.Type {
+		case sourceLabel:
+			// TODO
+		}
+	}
+	return diags
+}
+
+// RequiredPlugin represents a declaration of a dependency on a particular
+// Plugin version or source.
+type RequiredPlugin struct {
+	Name        string
+	Source      string
+	Type        *addrs.Plugin
+	Requirement VersionConstraint
+	DeclRange   hcl.Range
+}
+
+type RequiredPlugins struct {
+	RequiredPlugins map[string]*RequiredPlugin
+	DeclRange       hcl.Range
+}
+
+func decodeRequiredPluginsBlock(block *hcl.Block) (*RequiredPlugins, hcl.Diagnostics) {
+	attrs, diags := block.Body.JustAttributes()
+	ret := &RequiredPlugins{
+		RequiredPlugins: make(map[string]*RequiredPlugin),
+		DeclRange:       block.DefRange,
+	}
+	for name, attr := range attrs {
+		expr, err := attr.Expr.Value(nil)
+		if err != nil {
+			diags = append(diags, err...)
+		}
+
+		nameDiags := checkPluginNameNormalized(name, attr.Expr.Range())
+		diags = append(diags, nameDiags...)
```

Comment:
wouldn't be better to stop here in case diags has error?

Reply:
forget about it! I guess you want to parse the following and append any error

Reply:
> I guess you want to parse the following and append any error

Yup !